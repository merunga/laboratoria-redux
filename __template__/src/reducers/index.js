import { combineReducers } from 'redux'
export const initialState = {
  initialValue: 0,
}

export default combineReducers({
  secondValue: (state, action) => 2,
})
