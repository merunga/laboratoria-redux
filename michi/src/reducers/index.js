import { combineReducers } from 'redux'
import inState from './initialState'
import michi from './michi'

export const initialState = inState

export default combineReducers({
  michi,
})
