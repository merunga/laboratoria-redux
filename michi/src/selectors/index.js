const countCells = (michi) => {
  const accum = {
    X: 0,
    O: 0,
    [null]: 0,
  }
  const row = [0, 1, 2]
  /* eslint-disable array-callback-return */
  row.map((i) => {
    row.map((j) => {
      const value = michi[i][j]
      accum[value] += 1
    })
  })
  /* eslint-enable array-callback-return */
  return accum
}

export const turnoActual = ({ michi }) => {
  const accum = countCells(michi)
  return accum.X > accum.O ? 'O' : 'X'
}

const ganadoFila = (michi) => {
  for (let i = 0; i < 3; i += 1) {
    if (michi[i][0] && michi[i][0] === michi[i][1] && michi[i][0] === michi[i][2]) {
      return {
        jugador: michi[i][0],
        jugada: [
          { x: i, y: 0 },
          { x: i, y: 1 },
          { x: i, y: 2 },
        ],
      }
    }
  }
  return { jugador: null, jugada: null }
}

const ganadorColumna = (michi) => {
  for (let j = 0; j < 3; j += 1) {
    if (michi[0][j] && michi[0][j] === michi[1][j] && michi[0][j] === michi[2][j]) {
      return {
        jugador: michi[0][j],
        jugada: [
          { x: 0, y: j },
          { x: 1, y: j },
          { x: 2, y: j },
        ],
      }
    }
  }
  return { jugador: null, jugada: null }
}

const ganadorDiagonal = (michi) => {
  if (michi[0][0] && michi[0][0] === michi[1][1] && michi[0][0] === michi[2][2]) {
    return {
      jugador: michi[0][0],
      jugada: [
        { x: 0, y: 0 },
        { x: 1, y: 1 },
        { x: 2, y: 2 },
      ],
    }
  }
  if (michi[2][0] && michi[2][0] === michi[1][1] && michi[2][0] === michi[0][2]) {
    return {
      jugador: michi[2][0],
      jugada: [
        { x: 2, y: 0 },
        { x: 1, y: 1 },
        { x: 0, y: 2 },
      ],
    }
  }
  return { jugador: null, jugada: null }
}

export const ganador = ({ michi }) => (
  ganadoFila(michi).jugador ||
  ganadorColumna(michi).jugador ||
  ganadorDiagonal(michi).jugador
)

export const empate = ({ michi }) => {
  if (ganador({ michi })) {
    return false
  }
  return countCells(michi).null === 0
}

export const juegoTerminado = ({ michi }) => !!(ganador({ michi }) || empate({ michi }))

export const jugadaGanadora = ({ michi }) => (
  ganadoFila(michi).jugada ||
  ganadorColumna(michi).jugada ||
  ganadorDiagonal(michi).jugada
)

// export const turnoCount = ({ michi }) => 0
