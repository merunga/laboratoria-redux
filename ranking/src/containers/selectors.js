export const canciones = state => state.canciones.sort((c1, c2) => c1.votos <= c2.votos)

export const seleccionado = state => state.canciones.find(c => c.id === state.seleccionado) || null
