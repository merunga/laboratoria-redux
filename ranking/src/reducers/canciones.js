import { actionTypes } from '../actions'

export default (state, action) => {
  if (action.type === actionTypes.UP_VOTE) {
    return state.map(c => (c.id === action.id ? { ...c, votos: c.votos + 1 } : c))
  }
  if (action.type === actionTypes.DOWN_VOTE) {
    return state.map(c => (
      c.id === action.id ? { ...c, votos: (c.votos > 0 ? c.votos - 1 : 0) } : c
    ))
  }
  return state || null
}
