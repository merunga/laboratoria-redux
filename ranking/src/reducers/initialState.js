export default {
  canciones: [
    {
      id: 1, titulo: 'Psycho', autor: 'Post Malone', votos: 70,
    },
    {
      id: 2, titulo: 'Perfect', autor: 'Ed Sheeran', votos: 90,
    },
    {
      id: 3, titulo: 'God\'s Plan', autor: 'Drake', votos: 100,
    },
    {
      id: 4, titulo: 'Finesse', autor: 'Bruno Mars', votos: 80,
    },
    {
      id: 5, titulo: 'Meant To Be', autor: 'Bebe Rexha & Florida Georgia Line', votos: 60,
    },
  ],
  seleccionado: 1,
}
