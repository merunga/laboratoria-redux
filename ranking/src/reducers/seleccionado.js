import { actionTypes } from '../actions'

export default (state, action) => {
  switch (action.type) {
    case actionTypes.PREV:
    case actionTypes.NEXT: {
      const { canciones } = action
      const seleccionado = canciones.find(c => c.id === state) || null
      let selIdx = canciones.indexOf(seleccionado)
      switch (action.type) {
        case actionTypes.PREV:
          selIdx = selIdx > 0 ? selIdx - 1 : canciones.length - 1
          break
        case actionTypes.NEXT:
          selIdx = selIdx < canciones.length - 1 ? selIdx + 1 : 0
          break
        default:
          break
      }
      return canciones[selIdx].id
    }
    case actionTypes.SELECT:
      return action.id
    default:
      break
  }
  return state || null
}
