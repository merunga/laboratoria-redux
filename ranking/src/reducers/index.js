import { combineReducers } from 'redux'
import inState from './initialState'
import canciones from './canciones'
import seleccionado from './seleccionado'

export const initialState = inState

export default combineReducers({
  canciones,
  seleccionado,
})
