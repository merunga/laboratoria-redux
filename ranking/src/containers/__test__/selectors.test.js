import * as selectors from '../selectors'
import { initialState } from '../../reducers'

describe('selectors', () => {
  it('canciones', () => {
    expect(selectors.canciones(initialState)).toMatchSnapshot()
  })
  it('seleccionado', () => {
    expect(selectors.seleccionado(initialState)).toMatchSnapshot()
  })
})
