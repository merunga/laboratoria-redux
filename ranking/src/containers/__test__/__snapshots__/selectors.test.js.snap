// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors canciones 1`] = `
Array [
  Object {
    "autor": "Drake",
    "id": 3,
    "titulo": "God's Plan",
    "votos": 100,
  },
  Object {
    "autor": "Ed Sheeran",
    "id": 2,
    "titulo": "Perfect",
    "votos": 90,
  },
  Object {
    "autor": "Bruno Mars",
    "id": 4,
    "titulo": "Finesse",
    "votos": 80,
  },
  Object {
    "autor": "Post Malone",
    "id": 1,
    "titulo": "Psycho",
    "votos": 70,
  },
  Object {
    "autor": "Bebe Rexha & Florida Georgia Line",
    "id": 5,
    "titulo": "Meant To Be",
    "votos": 60,
  },
]
`;

exports[`selectors seleccionado 1`] = `
Object {
  "autor": "Post Malone",
  "id": 1,
  "titulo": "Psycho",
  "votos": 70,
}
`;
